addons/docs/src/frameworks/common/enhanceSource.ts

```diff
@@ -0,0 +1,53 @@
+import { ParameterEnhancer, combineParameters } from '@storybook/client-api';
+
+interface Location {
+  line: number;
+  col: number;
+}
+
+interface StorySource {
+  source: string;
+  locationsMap: { [id: string]: { startBody: Location; endBody: Location } };
+}
+
+const extract = (targetId: string, { source, locationsMap }: StorySource) => {
+  if (!locationsMap) {
+    return source;
+  }
+  const location = locationsMap[targetId];
+
+  // FIXME: bad locationsMap generated for module export functions whose titles are overridden
+  if (!location) return null;
+  const { startBody: start, endBody: end } = location;
+  const lines = source.split('\n');
+  if (start.line === end.line && lines[start.line - 1] !== undefined) {
+    return lines[start.line - 1].substring(start.col, end.col);
+  }
+  // NOTE: storysource locations are 1-based not 0-based!
+  const startLine = lines[start.line - 1];
+  const endLine = lines[end.line - 1];
+  if (startLine === undefined || endLine === undefined) {
+    return source;
+  }
+  return [
+    startLine.substring(start.col),
+    ...lines.slice(start.line, end.line - 1),
+    endLine.substring(0, end.col),
+  ].join('\n');
+};
+
+export const enhanceSource: ParameterEnhancer = context => {
+  const { id, parameters } = context;
+  const { storySource, docs = {} } = parameters;
+  const { formatSource, source } = docs;
+
+  // no input or user has manually overridden the output
+  if (!storySource?.source || source?.code) {
```

Comment:
Are we using chaining? this is a ESNext feature, not supported by node even in 12

Reply:
```suggestion
  if (!storySource?.source || docs.source?.code) {
```
I think this is easier to read, otherwise all the `source`s are pretty confusing.

Reply:
This TypeScript tho, and chaining is supported in TypeScript. I think this (and other parameters code) could really benefit from it.

Reply:
Sure